rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+
+1. `x` has type `T`
+2. `y` has the same type as `x`
+
+But when `x` is used as the conditional in an if statement, we can add an additional constraint:
+
+3. the type of `x` is `bool`
+
+With these three constraints, the toy compiler is able to perform constraint solving (unifying the constraints with unification) to determine that both `x` and `y` are of type `bool`.
+
+Now, imagine that the toy language includes addition expressions and the toy compiler encounters an addition expression `y + 1u64`. This would generate the constraint:
+
+4. `y` has type `u64`
+
+With these four constraints, the toy compiler is unable to perform unification, as the constraints are unable to be unified and generate a type error.
+
+We can see that the toy compiler would be able to perform type inference for a whole program by generating a list of constraints and then unifying these constraints.
+
+(Type inference, constraint solving, and unification are defined here: https://papl.cs.brown.edu/2014/Type_Inference.html)
+
+## Declaration Engine
+
+The declaration engine takes the concept of constraints and unification and applies it to declarations.
+
+For example, when type checking a function application, the compiler does not need to care about the contents of the body of the function, instead it is sufficient to create a constraint referring to "the return type of the function".
+
+For example, given a typed function declaration `add`, we add this declaration to the declaration engine. Then when type checking a function application of `add`, instead of inlining the function body, we create constraints referring to "the body of `add`" and "the return type of `add`". When type checking a _function application_
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This design will take after the design of the [type engine](https://github.com/FuelLabs/sway/blob/2816c13698a35752136d4843dbfe5e1b95c26e10/sway-core/src/type_engine/engine.rs#L14). During type checking, the compiler will add information about declarations to the declaration engine. Then, in places where that declaration is used, a value representing that declaration can be generated in place of inlining the declaration.
+
+### Definition
+
+During type checking of a declaration, the compiler will add information about that declaration to the declaration engine:
+
+```rust
+struct DeclarationEngine {
+    // function name to typed function declaration
+    // this example code does not consider Path's, but we could use
+    // a BTreeMap with Path's as the key
+    functions: HashMap<String, TypedFunctionDeclaration>,
+
+    // struct name to typed struct declaration
+    // again, this does not consider Path's
+    structs: HashMap<String, TypedStructDeclaration>,
+
+    // enums and traits
+}
+
+impl DeclarationEngine {
+    fn add_function(&mut self, function: TypedFunctionDeclaration) -> DeclarationDefinition {
```

Comment:
If the function has to be typed when we insert it into the declaration, doesn't that mean we maintain the status quo of having to type check (and perform inference on) everything before beginning this process?